consensus/packets/tristatebitset.go

```diff
@@ -60,13 +63,151 @@ func (dbs *TriStateBitSet) ApplyChanges(changes []BitSetCell) {
 }
 
 func (dbs *TriStateBitSet) Serialize() ([]byte, error) {
-	return nil, nil
+	var firstByte uint8
+	var secondByte uint8
+	if dbs.CompressedSet {
+		firstByte = 0x01
+	} else {
+		firstByte = 0x00
+	}
+
+	array, err := dbs.cellsToBitArray()
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to get bitarray from cells")
+	}
+
+	tmpLen := array.Len()
+	totalSize := int(math.Round(float64((tmpLen*2)/8))+0.5) + 1 // first byte
+	var result *bytes.Buffer
+	firstByte = firstByte << 1
+	if bits.Len(tmpLen) > 6 {
+		firstByte++
+		firstByte = firstByte << 6
+		secondByte = uint8(tmpLen)
+		totalSize++ // secondbyte is optional
+		result = allocateBuffer(totalSize)
+		err = binary.Write(result, defaultByteOrder, firstByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+		err = binary.Write(result, defaultByteOrder, secondByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+	} else {
+		result = allocateBuffer(totalSize)
+		firstByte = firstByte << 1
+		firstByte += uint8(tmpLen)
+		err = binary.Write(result, defaultByteOrder, firstByte)
+		if err != nil {
+			return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+		}
+	}
+
+	data, err := array.serialize()
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to serialize a bitarray")
+	}
+	err = binary.Write(result, defaultByteOrder, data)
+	if err != nil {
+		return nil, errors.Wrap(err, "[ Serialize ] failed to write binary")
+	}
+
+	return result.Bytes(), nil
 }
 
 func (dbs *TriStateBitSet) Deserialize(data io.Reader) error {
+	var firstbyte uint8
+	err := binary.Read(data, defaultByteOrder, &firstbyte)
+	if err != nil {
+		return errors.Wrap(err, "[ Deserialize ] failed to read first byte")
+	}
+	compressed, hbitFlag, length := parseFirstByte(firstbyte)
+	if hbitFlag {
+		err = binary.Read(data, defaultByteOrder, &length)
+		if err != nil {
+			return errors.Wrap(err, "[ Deserialize ] failed to read second byte")
+		}
+	}
+	blockCount := uint64(math.Round(float64(length/8) + 0.5))
+	payload := make([]uint8, blockCount)
+	for i := 0; uint64(i) < blockCount; i++ {
+		err = binary.Read(data, defaultByteOrder, &payload[i])
+		if err != nil {
+			return errors.Wrap(err, "[ Deserialize ] failed to read first byte")
+		}
+	}
+	var array *bitArray
+	if compressed {
+		// TODO: uncompress payload
```

Comment:
add panic here whilst we have no implementation for this branch

Reply:
done